File: problem-solving/lc-375-medium-guess-number-higher-or-lower-ii/2-bottom-up.go
/*
https://leetcode.com/problems/guess-number-higher-or-lower-ii/

Complexity
----------
Time: O(n^3)
Space: O(n^2)
*/

func getMoneyAmount(n int) int {
    var (
        dp [][]int
        temp int
    )
    for i := 0; i < n + 1; i++ {
        dp = append(dp, make([]int, n + 1))
    }

    // This loop in run in reverse order because in the recurrence relation,
    // we are accessing future index (g + 1) of l.
    for l := n - 1; l > 0; l-- {
        for h := l + 1; h <= n; h++ {
            dp[l][h] = math.MaxInt32
            for g := l; g < h; g++ {
                temp = g + max(dp[l][g - 1], dp[g + 1][h])
                dp[l][h] = min(dp[l][h], temp)
            }
        }
    }
    return dp[1][n]
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: problem-solving/lc-1155-medium-number-of-dice-rolls-with-target-sum/2-bottom-up.py
"""
https://leetcode.com/problems/number-of-dice-rolls-with-target-sum/

Complexity
----------
Time: O(n*k*target)
Space: O(n*target)
"""

class Solution:
    def numRollsToTarget(self, n: int, k: int, target: int) -> int:
        dp = [[0] * (target + 1) for _ in range(n + 1)]
        dp[0][0] = 1
        for i in range(1, n + 1):
            for tg in range(1, target + 1):
                for x in range(1, min(k, tg) + 1):
                    dp[i][tg] = dp[i][tg] + dp[i - 1][tg - x]
        return dp[n][target] % (10**9 + 7)
